crates/bevy_ecs/src/event/base.rs

```diff
@@ -48,8 +47,41 @@ pub trait Event: Component {
     /// [triggered]: crate::system::Commands::trigger_targets
     /// [`Trigger::propagate`]: crate::observer::Trigger::propagate
     const AUTO_PROPAGATE: bool = false;
+
+    /// Generates the [`ComponentId`] for this event type.
+    ///
+    /// If this type has already been registered,
+    /// this will return the existing [`ComponentId`].
+    ///
+    /// This is used by various dynamically typed observer APIs.
+    fn register_component_id(world: &mut World) -> ComponentId {
+        world.register_component::<EventWrapperComponent<Self>>()
+    }
+
+    /// Fetches the [`ComponentId`] for this event type,
+    /// if it has already been generated.
+    ///
+    /// This is used by various dynamically typed observer APIs.
+    fn get_component_id(world: &World) -> Option<ComponentId> {
+        world.component_id::<EventWrapperComponent<Self>>()
+    }
 }
 
+/// An internal type that implements [`Component`] for a given [`Event`] type.
+///
+/// This exists so we can easily get access to a unique [`ComponentId`] for each [`Event`] type,
+/// without requiring that [`Event`] types implement [`Component`] directly.
+/// [`ComponentId`] is used internally as a unique identitifier for events because they are:
+///
+/// - Unique to each event type.
+/// - Can be quickly generated and looked up.
+/// - Are compatible with dynamic event types, which aren't backed by a Rust type.
+///
+/// This type is an implementation detail and should never be made public.
+// TODO: refactor events to store their metadata on distinct entities, rather than using `ComponentId`
+#[derive(Component)]
+struct EventWrapperComponent<E: Event + ?Sized>(PhantomData<E>);
```

Comment:
Note that doing it this way won't give usable values for `layout` or `drop` in the `ComponentInfo`.  I don't think anyone is actually using those right now, but the reason I originally suggested a `#[repr(transparent)]` wrapper instead of `PhantomData` is that those values would be valid for the event type.  

```suggestion
#[repr(transparent)]
struct EventWrapperComponent<E: Event>(E);
```